File: packages/python/docparse/async_client.py
"""Async job client for large-document extraction via Mawlaia DocParse API."""
from __future__ import annotations
import time
from typing import Optional
import httpx

_DEFAULT_BASE = "https://api.mawlaia.com"


class AsyncJobClient:
    """
    Submit DocParse extraction jobs and poll for results.
    Use this for large PDFs (>1000 words) where the sync /extract endpoint may time out.

    Example::

        from docparse import AsyncJobClient

        client = AsyncJobClient(api_key="mwl_live_...")

        job = client.submit(text=pdf_text, schema_name="contract",
                            webhook_url="https://yourapp.com/hooks/docparse")
        print(job["job_id"], job["status"])  # pending

        # Poll until done (or use webhook)
        result = client.wait(job["job_id"], timeout=120)
        print(result["result"])  # {"parties": {"value": [...], "confidence": 0.97}, ...}
    """

    def __init__(self, api_key: str, base_url: str = _DEFAULT_BASE, timeout: float = 15.0):
        self._base    = base_url.rstrip("/")
        self._headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
        self._timeout = timeout

    def submit(
        self,
        text:        str,
        schema_name: Optional[str] = None,
        fields:      Optional[list[dict]] = None,
        webhook_url: Optional[str] = None,
    ) -> dict:
        """Submit a job. Returns immediately with ``{"job_id": ..., "status": "pending"}``."""
        payload: dict = {"text": text}
        if schema_name:
            payload["schema_name"] = schema_name
        if fields:
            payload["fields"] = fields
        if webhook_url:
            payload["webhook_url"] = webhook_url

        with httpx.Client(base_url=self._base, headers=self._headers, timeout=self._timeout) as c:
            r = c.post("/v1/doc/jobs", json=payload)
            r.raise_for_status()
            return r.json()

    def poll(self, job_id: str) -> dict:
        """Fetch current status of a job."""
        with httpx.Client(base_url=self._base, headers=self._headers, timeout=self._timeout) as c:
            r = c.get(f"/v1/doc/jobs/{job_id}")
            r.raise_for_status()
            return r.json()

    def list_jobs(self) -> list[dict]:
        """Return the 20 most recent jobs for this API key."""
        with httpx.Client(base_url=self._base, headers=self._headers, timeout=self._timeout) as c:
            r = c.get("/v1/doc/jobs")
            r.raise_for_status()
            return r.json()

    def wait(self, job_id: str, timeout: float = 120.0, poll_interval: float = 2.0) -> dict:
        """
        Block until the job is ``completed`` or ``failed``, then return the job dict.
        Raises ``TimeoutError`` if *timeout* seconds elapse first.
        """
        deadline = time.monotonic() + timeout
        while True:
            job = self.poll(job_id)
            if job["status"] in ("completed", "failed"):
                return job
            if time.monotonic() >= deadline:
                raise TimeoutError(f"Job {job_id} not finished within {timeout}s")
            time.sleep(poll_interval)

File: packages/typescript/src/async_client.ts
export interface AsyncJobClientOptions {
  apiKey: string;
  baseUrl?: string;
  timeout?: number;
}

export interface DocparseJob {
  job_id: string;
  status: "pending" | "processing" | "completed" | "failed";
  schema_name?: string;
  result?: Record<string, unknown>;
  error?: string;
  created_at: string;
  completed_at?: string;
}

export class AsyncJobClient {
  private readonly base: string;
  private readonly headers: Record<string, string>;
  private readonly timeout: number;

  constructor({ apiKey, baseUrl = "https://api.mawlaia.com", timeout = 15_000 }: AsyncJobClientOptions) {
    this.base = baseUrl.replace(/\/$/, "");
    this.headers = {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    };
    this.timeout = timeout;
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), this.timeout);
    try {
      const res = await fetch(`${this.base}${path}`, {
        method,
        headers: this.headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });
      if (!res.ok) throw new Error(`Mawlaia API error: ${res.status} ${await res.text()}`);
      return res.json() as Promise<T>;
    } finally {
      clearTimeout(id);
    }
  }

  async submit(
    text: string,
    options: { schemaName?: string; fields?: Record<string, unknown>[]; webhookUrl?: string } = {},
  ): Promise<{ job_id: string; status: "pending" }> {
    const body: Record<string, unknown> = { text };
    if (options.schemaName) body.schema_name = options.schemaName;
    if (options.fields)     body.fields      = options.fields;
    if (options.webhookUrl) body.webhook_url  = options.webhookUrl;
    return this.request("POST", "/v1/doc/jobs", body);
  }

  async poll(jobId: string): Promise<DocparseJob> {
    return this.request("GET", `/v1/doc/jobs/${jobId}`);
  }

  async listJobs(): Promise<DocparseJob[]> {
    return this.request("GET", "/v1/doc/jobs");
  }

  async wait(jobId: string, timeoutMs = 120_000, pollIntervalMs = 2_000): Promise<DocparseJob> {
    const deadline = Date.now() + timeoutMs;
    while (true) {
      const job = await this.poll(jobId);
      if (job.status === "completed" || job.status === "failed") return job;
      if (Date.now() >= deadline) throw new Error(`Job ${jobId} not finished within ${timeoutMs}ms`);
      await new Promise((r) => setTimeout(r, pollIntervalMs));
    }
  }
}
